lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,193 @@
+    FIELD_TIMING_CALLBACKS_KEY = :field_timing_callbacks
+    ANALYZER_INSTANCE_KEY = :analyzer_instance
+
+    # NOTE: These constants are used both to match tracing keys from graphql-ruby as well as to store in-line and lazy
+    # field resolution timings in context.
+    GRAPHQL_TRACING_FIELD_KEY = 'execute_field'
+    GRAPHQL_TRACING_LAZY_FIELD_KEY = 'execute_field_lazy'
+
+    class << self
+      def current_time
+        Process.clock_gettime(Process::CLOCK_REALTIME)
+      end
+
+      def current_time_monotonic
+        Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      end
+
+      def before_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
+
+        query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[RUNTIME_METRICS_ENABLED] = true
+          ns[FIELD_TIMING_CALLBACKS_KEY] = []
+          ns[QUERY_START_TIME_KEY] = current_time
+          ns[QUERY_START_TIME_MONOTONIC_KEY] = current_time_monotonic
+          ns[GRAPHQL_TRACING_FIELD_KEY] = {}
+          ns[GRAPHQL_TRACING_LAZY_FIELD_KEY] = {}
+        end
+      end
+
+      def after_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
+
+        query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          duration = current_time_monotonic - ns[QUERY_START_TIME_MONOTONIC_KEY]
+          runtime_metrics = { start_time: ns[QUERY_START_TIME_KEY], duration: duration }
+
+          analyzer = ns[ANALYZER_INSTANCE_KEY]
+          analyzer.extract_query(runtime_metrics: runtime_metrics, context: query.context)
+          analyzer.run_field_timing_callbacks
+        end
+      end
+
+      def trace(key, data, &block)
+        skip_tracing = data[:query]&.context&.fetch(:skip_graphql_metrics_analysis, false)
+        return block.call if skip_tracing || ![GRAPHQL_TRACING_FIELD_KEY, GRAPHQL_TRACING_LAZY_FIELD_KEY].include?(key)
+        # NOTE: We can't just check `runtime_metrics_enabled?` here, since .trace runs before `before_query`, i.e.
+        # during lexing, parsing, validation etc., and `before_query` doesn't run until `execute_multiplex`.
+
+        if runtime_metrics_enabled?(data[:query]&.context)
+          trace_field(key, data, block, key)
+        else
+          block.call
+        end
+      end
+
+      def trace_field(_key, data, block, context_key)
+        path_excluding_numeric_indicies = data[:path].select { |p| p.is_a?(String) }
+
+        start_time = current_time
+        start_time_monotonic = current_time_monotonic
+
+        result = block.call
+        duration = current_time_monotonic - start_time_monotonic
+
+        data[:query].context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[context_key][path_excluding_numeric_indicies] ||= []
+          ns[context_key][path_excluding_numeric_indicies] << { start_time: start_time, duration: duration }
```

Comment:
does this mean we're collecting the same field referenced through different paths under different events? that seems like it would unnecessarily inflate our event count and make aggregation harder?

Reply:
It just means that a given field, resolved n times within a list, connection etc. would be attributed n (multiplied for nested lists/connections) timings.

So:

```graphql
products(first: 5) {
  edges {
    node {
      title
      variants(first: 2) {
        title
      }
    }
  }
}
```

We extract fields only once, but in this example we may have 5 timings attributed to `shop.products.edges.node.title`, and 10 timings attributed to `shop.products.edges.node.variants.edges.node.title`.